### PR TITLE
Reverse stretched mesh for land model

### LINKED ISSUE #995: Reverse stretched grids for land model
**Is your feature request related to a problem? Please describe.**
The atmospheric model uses a stretched grid (via exponential or generalized exponential stretching functions) for which the smallest grid spacing happens near the surface (called `dz_bottom` in the code and in ClimaAtmos CLI options) and the largest near the top of the atmosphere (called `dz_top`). 

However, the land model requires the opposite behavior. That is, the smallest grid spacing near the top surface (just at subsurface level) and largest grid spacing at the bottom (deep down towards the Earth core). 

**Describe the solution you'd like**

**Describe alternatives you've considered**

- For the `ExponentialStretching` function, which currently has only one parameter (the height scaling factor, H), I propose adding the optional boolean parameter `reverse_mode`, so that the typical user of the land model would specify

  `stretching = Meshes.ExponentialStretching(H; reverse_mode = true)` 

  Note that the boolean parameter (`false` by default) will align with the existing use-case for the atmosphere. So only users who want to reverse this function will have to set this to `true`. 

  Internally, we would just use the opposite `η(ζ)` function, i.e., instead of the current [`η(ζ)`](https://github.com/CliMA/ClimaCore.jl/blob/aa15859549770caf317d7ea786576758cdc26f96/src/Meshes/interval.jl#L147), defined as `η(ζ) = -h * log1p((expm1(-1 / h)) * ζ)`, we will use,  `η(ζ) = h * log1p((expm1(-1 / h)) * ζ)` and this change will be easily enabled by the boolean flag.

- The `GeneralizedExponentialStretching` requires a bit more thinking. 
  The nature of the signature of this function would not change, and I would still add an optional boolean parameter, i.e., 

  `stretching = Meshes.GeneralizedExponentialStretching(dz_bottom, dz_top; reverse_mode)`

  `dz_bottom` will still be specified as the first argument and `dz_top` as the second one. But internally, we have to account for the fact that what was the role of the smallest grid spacing for the atmosphere (`dz_bottom`) is now the role of `dz_top` and vice versa. 

**Additional context**
I added a section to the design-docs, section _3.3.3 Reversed exponential stretching (for land model)_ with the proposed model, but I would appreciate more guidance to know if I got off track (@tapios , @kmdeck , @charleskawczynski )

Reviewers: @kmdeck and @charleskawczynski

### PR DESCRIPTION
**Pull request by valeriabarra**
This will close #995

This PR adds a new command line argument for the creation of a stretched mesh, both with `ExponentialStretching` and `GeneralizedExponentialStretching` stretching types. 

To activate the reverse mode for land models, the user will have to pass a boolean flag (`false` by default, to reflect the existing mode for atmos configurations) to the stretching mode function. The new interface will look like: 

```stretching = Meshes.ExponentialStretching(H; reverse_mode)```

and 

```stretching = Meshes.GeneralizedExponentialStretching(dz_bottom, dz_top; reverse_mode)```

In addition to the new feature and new unit tests, this PR improves existing documentation and tests. In the `ExponentialStretching` unit tests, we were originally only checking the generated mesh faces against a reference profile. I left these checks and added checks without a reference profile, so that we effectively check that the transformation (and its inverse) give us the desired behavior. 

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: src/Meshes/interval.jl [hunk 1]
```diff
@@ -136,102 +138,211 @@ function IntervalMesh(
...
-    faces =
-        [CT(cmin + R * η(ζ)) for ζ in range(FT(0), FT(1); length = nelems + 1)]
+    reverse_coeff = reverse_mode ? -1 : 1
+    η(ζ) =
+        ζ == 1 ? reverse_coeff * ζ :
+        reverse_coeff * (-h) * log1p((expm1(-1 / h)) * ζ)
+    faces = [
+        CT(reverse_mode ? cmax + R * η(ζ) : cmin + R * η(ζ)) for
+        ζ in range(FT(0), FT(1); length = nelems + 1)
+    ]
```
Comment (charleskawczynski) [inline src/Meshes/interval.jl:161]: Can we get away without these changes?

**Review by charleskawczynski — COMMENTED** [review]
It might be nice to do a peel offf PR and change the `dz_surface` to `dz_bottom`, first. It's difficult to spot which changes are due to name changes only vs added feature

File: src/Meshes/interval.jl [hunk 2]
```diff
@@ -167,34 +189,54 @@ function IntervalMesh(
...
+    reverse_mode::Bool = false,
 ) where {CT <: Geometry.Abstract1DPoint{FT}} where {FT}
     if nelems ≤ 1
         throw(ArgumentError("`nelems` must be ≥ 2"))
     end
+
+    if nelems ≤ 1
+        throw(ArgumentError("`nelems` must be ≥ 2"))
+    end
+
```
Comment (charleskawczynski) [inline src/Meshes/interval.jl:201]:
```suggestion
```

File: test/Meshes/interval.jl [hunk 3]
```diff
@@ -126,6 +136,47 @@ end
...
+    )
+    mesh = unit_intervalmesh(
+        stretching = Meshes.ExponentialStretching(0.25),
+        nelems = 1,
+        reverse_mode = true,
+    )
+    @test Meshes.coordinates(mesh, 1, 1) == Geometry.ZPoint(-1.0)
+    @test Meshes.coordinates(mesh, 1, 2) ≈ Geometry.ZPoint(0.0)
+    # approx equal to gcm ref profile height fraction given the vert domain
+    # ~7.5 km of 45 km vertical domain extent
```
Comment (tapios) [inline test/Meshes/interval.jl:153]: =```suggestion
    # height of 45 km and scale height of ~7.5 km 
```

File: test/Meshes/interval.jl [hunk 4]
```diff
@@ -126,6 +136,47 @@ end
...
+    # we normalize to unit length for easy comparison
+    mesh = unit_intervalmesh(
+        stretching = Meshes.ExponentialStretching(7.5 / 45.0),
```
Comment (tapios) [inline test/Meshes/interval.jl:156]: Rather than hard coding a mesh for testing, it's better to test that the grid spacing is equidistant in exp(-z/H), where H is the scale height.

Comment (valeriabarra) [reply test/Meshes/interval.jl:156]: Oh, I followed the same test pattern that was already there for the non-reversed stretched grid. I am not sure I entirely follow the suggestion though. Can you elaborate a bit more? We have the separate tests for the `ExponentialStretch` which scales with the height scale H

File: src/Meshes/interval.jl (hunk 2 above)
Comment (valeriabarra) [reply src/Meshes/interval.jl:201]: Why would you remove this? `GeneralizedExponentialStretching` only works with two arguments that specify the bottom and top spacing so I can see why the original implementer thought about not allowing a mesh with fewer than 2 elements because you would not have a way to specify both bottom and top spacing with fewer than 2 elements

File: test/Meshes/interval.jl (hunk 4 above)
Comment (valeriabarra) [reply test/Meshes/interval.jl:156]: Oh, I think I got it now! Basically check that the inverse transformation gives ewuispaced points.

File: src/Meshes/interval.jl (hunk 2 above)
Comment (charleskawczynski) [reply src/Meshes/interval.jl:201]: It's a duplicate of lines directly above, it looks like a rebase error

Comment (valeriabarra) [reply src/Meshes/interval.jl:201]: Ah, thanks for the good catch. Yeah, it must have been from the rebase. I'll remove it and squash

**Comment by valeriabarra** [comment]
Thanks @charleskawczynski . I have now cleaned it up and tests (both existing and improved ones) all pass. If approved, I'll squash commits

File: src/Meshes/interval.jl [hunk 5]
```diff
@@ -167,34 +191,50 @@ function IntervalMesh(
...
     nelems::Int,
+    reverse_mode::Bool = false,
 ) where {CT <: Geometry.Abstract1DPoint{FT}} where {FT}
     if nelems ≤ 1
         throw(ArgumentError("`nelems` must be ≥ 2"))
     end
+
     dz_bottom, dz_top = stretch.dz_bottom, stretch.dz_top
-    if !(dz_bottom ≤ dz_top)
+    if !(dz_bottom ≤ dz_top) && reverse_mode == false
```
Comment (charleskawczynski) [inline src/Meshes/interval.jl:201]:
```suggestion
    if !(dz_bottom ≤ dz_top) && !reverse_mode
```
I think this reads a bit easier

File: src/Meshes/interval.jl [hunk 6]
```diff
@@ -167,34 +191,50 @@ function IntervalMesh(
...
         throw(ArgumentError("dz_bottom must be ≤ dz_top"))
     end
-    # surface coord height value
+
+    if !(dz_bottom ≥ dz_top) && reverse_mode == true
```
Comment (charleskawczynski) [inline src/Meshes/interval.jl:205]:
```suggestion
    if !(dz_bottom ≥ dz_top) && reverse_mode
```

File: src/Meshes/interval.jl [hunk 7]
```diff
@@ -167,34 +191,50 @@ function IntervalMesh(
...
     z_top = Geometry.component(domain.coord_max, 1)
+    # but in case of reverse_mode, we temporarily swap them so that the following root solve algorithm does not need to change
+    if reverse_mode
+        z_bottom, z_top = Geometry.component(domain.coord_max, 1),
+        -Geometry.component(domain.coord_min, 1)
+    end
+    # for the same reason, in case of reverse_mode, we swap the roles of dz_bottom (largest for land, smallest for atmos) and dz_top (largest for atmos, smallest for land)
+    if reverse_mode
+        dz_top, dz_bottom = dz_bottom, dz_top
+    end
```
Comment (charleskawczynski) [inline src/Meshes/interval.jl:221]:
```suggestion
    # but in case of reverse_mode, we temporarily swap them so that the following root solve algorithm does not need to change
    if reverse_mode
        z_bottom, z_top = Geometry.component(domain.coord_max, 1),
        -Geometry.component(domain.coord_min, 1)
        dz_top, dz_bottom = dz_bottom, dz_top
    end
```

File: src/Meshes/interval.jl (hunk 5 above)
Comment (valeriabarra) [reply src/Meshes/interval.jl]: It's subjective. For me that I have vision problem the version spelled out is easier to read than the one with the bang that could be easily missed at a first glance. But it's ok. I changed it as suggested in [0bf0a5a](https://github.com/CliMA/ClimaCore.jl/pull/1001/commits/0bf0a5ac9a4bc648a7796b7efd5465fe2d2beebf)

File: src/Meshes/interval.jl [hunk 8]
```diff
@@ -167,34 +191,48 @@ function IntervalMesh(
...
+    # so that the following root solve algorithm does not need to change
+    if reverse_mode
+        z_bottom, z_top = Geometry.component(domain.coord_max, 1),
+        -Geometry.component(domain.coord_min, 1)
+        dz_top, dz_bottom = dz_bottom, dz_top
+    end
 
     # define the inverse σ⁻¹ exponential stretching function
-    exp_stretch(ζ, h) = -h * log(1 - (1 - exp(-1 / h)) * ζ)
+    exp_stretch(ζ, h) = ζ == 1 ? ζ : -h * log(1 - (1 - exp(-1 / h)) * ζ)
```
Comment (charleskawczynski) [inline src/Meshes/interval.jl:222]: What is this change needed for? Is this to mirror what's done in the exponential stretching?

File: src/Meshes/interval.jl [hunk 9]
```diff
@@ -167,34 +191,48 @@ function IntervalMesh(
...
     nelems::Int,
+    reverse_mode::Bool = false,
 ) where {CT <: Geometry.Abstract1DPoint{FT}} where {FT}
     if nelems ≤ 1
         throw(ArgumentError("`nelems` must be ≥ 2"))
     end
+
     dz_bottom, dz_top = stretch.dz_bottom, stretch.dz_top
-    if !(dz_bottom ≤ dz_top)
+    if !(dz_bottom ≤ dz_top) && !reverse_mode
```
Comment (charleskawczynski) [inline src/Meshes/interval.jl:201]: Alternatively, we could leave this as

```julia
    if !(dz_bottom ≤ dz_top)
        throw(ArgumentError("dz_bottom must be ≤ dz_top"))
    end
```
and move it to below where `dz_bottom` and `dz_top` are recomputed, that may actually be simpler, what do you think?

File: src/Meshes/interval.jl (hunk 8 above)
Comment (valeriabarra) [reply src/Meshes/interval.jl:222]: Yes, because otherwise the corner case of the extremum might have not be exactly computed

**Review by charleskawczynski — APPROVED** [review]
